context/src/meta/validator.rs

```diff
@@ -0,0 +1,290 @@
+#[cfg(feature = "pyo3")]
+use pyo3::prelude::*;
+#[cfg(feature = "pyo3")]
+use pyo3_stub_gen::derive::{gen_stub_pyclass, gen_stub_pyclass_enum, gen_stub_pymethods};
+use thiserror::Error;
+#[cfg(feature = "wasm")]
+use wasm_bindgen::prelude::*;
+
+use crate::{
+    env::{
+        parser::BranchClass,
+        validator::{MAX_BRANCH_NAME_LEN, MAX_EMAIL_LEN, MAX_FIELD_LEN},
+    },
+    string_safety::{optional_string_to_empty_str, validate_field_len, FieldLen},
+};
+
+use super::MetaContext;
+
+#[cfg_attr(feature = "pyo3", gen_stub_pyclass_enum, pyclass(eq, eq_int))]
+#[cfg_attr(feature = "wasm", wasm_bindgen)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum MetaValidationLevel {
+    Valid = 0,
+    SubOptimal = 1,
+    Invalid = 2,
+}
+
+impl Default for MetaValidationLevel {
+    fn default() -> Self {
+        Self::Valid
+    }
+}
+
+impl ToString for MetaValidationLevel {
+    fn to_string(&self) -> String {
+        match self {
+            MetaValidationLevel::Valid => "VALID".to_string(),
+            MetaValidationLevel::SubOptimal => "SUBOPTIMAL".to_string(),
+            MetaValidationLevel::Invalid => "INVALID".to_string(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum MetaValidationIssue {
+    SubOptimal(MetaValidationIssueSubOptimal),
+    Invalid(MetaValidationIssueInvalid),
+}
+
+impl ToString for MetaValidationIssue {
+    fn to_string(&self) -> String {
+        match self {
+            Self::SubOptimal(i) => i.to_string(),
+            Self::Invalid(i) => i.to_string(),
+        }
+    }
+}
+
+#[derive(Error, Debug, Clone, PartialEq, Eq)]
+pub enum MetaValidationIssueInvalid {
+    #[error("CI info branch name too short")]
+    CIInfoBranchNameTooShort(String),
+    #[error("CI info is classified as a PR, but has not PR number")]
+    CIInfoPRNumberMissing,
+    #[error("CI info has a PR number, but branch is not classified as a PR")]
+    CIInfoPRNumberConflictsWithBranchClass,
+}
+
+#[derive(Error, Debug, Clone, PartialEq, Eq)]
+pub enum MetaValidationIssueSubOptimal {
```

Comment:
If I'm reading this right, this is a subset of `EnvValidationIssueSubOptimal`, as is `MetaValidationIssueInvalid` above. Can we refactor this a bit to avoid the duplication here?